docs/advanced/registration-sources.rst

```diff
@@ -20,6 +20,42 @@ You can use the ``Autofac.Features.ResolveAnything.AnyConcreteTypeNotAlreadyRegi
     builder.RegisterSource(new AnyConcreteTypeNotAlreadyRegisteredSource());
     var container = builder.Build();
 
+Contravariant Registration Source
+=================================
+The ``ContravariantRegistrationSource`` is helpful in registering types that need to be later resolved in a `contravariant context <https://docs.microsoft.com/en-us/dotnet/standard/generics/covariance-and-contravariance>`_ (using a more generic / less derived type than originally specified). This is common in handler patterns:
+
+.. sourcecode:: csharp
+
+    public interface IHandler<in TCommand>
+    {
+      void Handle(TCommand command);
+    }
+
+    public class CommandAHandler : IHandler<CommandA>
+    {
+      public void Handle(CommandA command)
+      {
+      }
+    }
+
+    public class CommandA
+    {
+    }
+
+    public class CommandB : CommandA
+    {
+    }
+
+    var builder = new ContainerBuilder();
+    builder.RegisterSource(new ContravariantRegistrationSource());
+    builder.RegisterType<CommandAHandler>().As<IHandler<CommandA>>();
+    var container = builder.Build();
+
+    // a and b are both CommandAHandler.
+    var a = container.Resolve<IHandler<CommandA>>();
+    var b = container.Resolve<IHandler<CommandB>>();
```

Comment:
Might I propose an example slightly closer to contravariance names, and using a similar Handle method to the example in the linked MS doc?
```suggestion
    public class BaseCommandHandler : IHandler<BaseCommand>
    {
      public void Handle(BaseCommand command)
      {
         Console.WriteLine(command.GetType().Name);
      }
    }

    public class BaseCommand
    {
    }

    public class DerivedCommand : BaseCommand
    {
    }

    var builder = new ContainerBuilder();
    builder.RegisterSource(new ContravariantRegistrationSource());
    builder.RegisterType<BaseCommandHandler>().As<IHandler<BaseCommand>>();
    var container = builder.Build();

    // a and b are both CommandHandler.
    var a = container.Resolve<IHandler<BaseCommand>>();
    a.Handle(new BaseCommand()); // prints BaseCommand
    var b = container.Resolve<IHandler<DerivedCommand>>();
    b.Handle(new DerivedCommand()); // prints DerivedCommand
```

Reply:
Sounds good to me. Committed. I wasn't sure how large or small the example should be. Somewhat debated not putting an example and referring people to the unit tests. ;)

Reply:
Cruel, but not unusual...